src/legacy/status_im/communities/e2e.cljs

```diff
@@ -1,35 +1,196 @@
+                     :chatIds       [channel-id]
+                     :tokenCriteria [((token-criteria token-symbol) amount)]}]})))
+
+(defn channel-token-criteria->requests
+  [community-id channel-id tks]
+  (let [keep-fn (partial channel-token-criteria->request community-id channel-id)]
+    (keep keep-fn tks)))
+
+(defn create-community-channel!
+  [channel community-id]
+  (->
+    (rpc/call
+     {:method      "wakuext_createCommunityChannel"
+      :js-response true
+      :params      [{:name        (:name channel)
+                     :communityId community-id
+                     :description (:name channel)}]
+      :on-error    #(log/error "failed to create token gated community channel."
+                               {:error        %
+                                :channel-name (:name channel)})})
+    (.then (fn [response]
+             (let [channel-id (js-response->channel-id (:name channel) response)]
+               (apply
+                p/chain
+                (p/resolved nil)
+                (channel-token-criteria->requests community-id channel-id (:permissions channel))))))))
+
+(defn create-community!
+  [community-name]
+  (rpc/call
+   {:method      "wakuext_createCommunity"
+    :js-response true
+    :params      [{:name                         community-name
+                   :description                  community-name
+                   :color                        "#887af9"
+                   :historyArchiveSupportEnabled true
+                   :membership                   constants/community-permissions-manual-accept
+                   :pinMessageAllMembersEnabled  true}]}))
+
+(defn create-token-gated-permission!
+  [{:keys [token-symbol permission-type amount-in-wei]} community-id]
+  (rpc/call
+   {:method      "wakuext_createCommunityTokenPermission2"
+    :js-response true
+    :params      [{:communityId   community-id
+                   :type          permission-type
+                   :tokenCriteria [((token-criteria token-symbol) amount-in-wei)]}]}))
+
+(defn create-community-from-description
+  [{:keys [community-name
+           community-permissions
+           channel-list]}]
+  (let [create-community-permissions-fn (map
+                                         #(response->community-id-fn
+                                           (partial create-token-gated-permission! %))
+                                         community-permissions)
+        create-channels-fn              (map
+                                         #(response->community-id-fn
+                                           (partial create-community-channel! %))
+                                         channel-list)]
+    (as-> (create-community! community-name) $
+      (apply p/chain $ create-community-permissions-fn)
+      (apply p/chain $ create-channels-fn)
+      (p/then $ #(rf/dispatch [:hide-bottom-sheet]))
+      (p/catch $ #(log/error "failed to create community e2e" {:error %})))))
```

Comment:
With funcitons returning functions, and also taking functions I think the logic is a bit hard to track, especially in those `map`s.

Please consider removing the `response->community-id-fn` function wrapper, and I'm wondering if it's possible to change `map` + `apply p/chian` -> `doseq`.

Not needed to change anything, it'd be also good to know if others can understand this well, we weren't too used to promises before, so that might also be a reason.

Reply:
So, we can use `doseq` etc, but we will have to nest I believe and capture the return value of the first promise, extract the community-id and pass it over, and i wanted to avoid nesting promises as much as possible.

The way it works is that any RPC in status-go, returns the same format (messenger-response), so chain works nicely with it as it's effectively creating a pipeline where each promise/rpc-output becomes the input of the next function, doseq/run! does't allow it as I think it discards the previous output in promesa , so the code would look I believe something like (pseudo code)

```
(-> (create-community ...)
     (.then (fn [response]
                  (let [community-id]
                     (p/run! 
                        #(create-community-permissions-fn community-id %)
                         createchannels))
     (p.then #(rf/dispatch....)) // this will have a nil promise, so we can't use the return value in case we need it, not an issue though)
)
```

I think the end result is conceptually a bit harder to grasp (i.e nesting promises rather than a pipeline of rpc calls), so more difficult to reason about if explained in words, but code wise it's probably easier to understand what's going on with the `doseq/run!` version.
Either are fine by me to be honest :)

Reply:
Hey @cammellos 

As we already talked about, I ended up thinking using `chain` is more concise, and actually this is a nice use of it :+1: 

I'd only ask for the simplification of the `let` bindings, I've tried to do it, so I think for:
```clojure
(map
 #(response->community-id-fn
   (partial create-token-gated-permission! %))
 community-permissions)
```

its equivalent code is (please confirm):
```clojure
(map (fn [permission]
       (fn [response]
         (create-token-gated-permission! permission (js-response->community-id response))))
     community-permissions)
```

I think this code is more explicit since we remove some abstractions, and we can clearly see what is the main function invoked and that this `map` will return a seq of fns, at the cost ofc of being more verbose.

Reply:
thanks for the suggestion, applied!